### PR TITLE
docs(release): script to generate a CHANGELOG.md

### PR DESCRIPTION
**Pull request by dpgaspar**
### SUMMARY
Helper script to help on Apache releases. Currently generating a change log is a pain. The script will output a `CHANGELOG` between two versions (tags or branches), may take a while because we want to map a git author to a github login and github has aggressive rate-limits

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: RELEASING/changelog.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+class GitChangeLog:
+    """
+    Helper class to output a list of logs entries on a superset changelog format
+
+    We want to map a git author to a github login, for that we call github's API
+    """
+
+    def __init__(self, logs: List[GitLog]) -> None:
+        self._logs = logs
+        self._github_login_cache = {}
```
Comment (dpgaspar) [inline RELEASING/changelog.py:65]: We can preset the cache here, for example with the currently more frequent committers

File: RELEASING/changelog.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+import json
+import os
+import re
+from dataclasses import dataclass
+from time import sleep
+from typing import Any, Dict, List, Optional, Union
+from urllib import request
+from urllib.error import HTTPError
+
+import click
```
Comment (mistercrunch) [inline RELEASING/changelog.py:26]: hmmm, probably don't want both line 26 and 29

**Comment by mistercrunch** [comment]
The aggressive rate limits is only if you don't have a github token specified, many of us have GITHUB_TOKEN env var defined in our bash_profile. Some libs take care of that but wouldn't sweat it.

**Comment by dpgaspar** [comment]
Trying to keep these scripts free of new dependencies. If it starts giving us problems we can add a specific github lib or prefill the cache

File: RELEASING/changelog.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+try:
+    import click
+except ModuleNotFoundError:
+    exit("Click is a required dependency for this script")
```
Comment (villebro) [inline RELEASING/changelog.py:31]: As `click` is a requirement in `requirements/base.txt`, can't we just assume this is installed?

File: RELEASING/changelog.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            os.popen('git --no-pager log --pretty=format:"%h|%an|%ad|%s|"')
+            .read()
+            .split("\n")
+        )
+
+    def _parse_log(self, log_item: str) -> GitLog:
+        pr_number = None
+        split_log_item = log_item.split("|")
+        # parse the PR number from the log message
+        match = re.match(".*\(\#(.*)\)", split_log_item[3])
```
Comment (villebro) [inline RELEASING/changelog.py:178]: nit:
```suggestion
        match = re.match(".*\(\#(\d*)\)", split_log_item[3])
```

File: RELEASING/changelog.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    def _wait_github_rate_limit(self) -> None:
+        """
+        Waits for available rate limit slots on the github API
+        """
+        while True:
+            rate_limit_payload = self._fetch_github_rate_limit()
+            if rate_limit_payload["rate"]["remaining"] > 1:
+                break
+            print(f".", end="", flush=True)
```
Comment (villebro) [inline RELEASING/changelog.py:76]: Nit:
```suggestion
            print(".", end="", flush=True)
```
We should probably also add a `print()` after the `while` to not continue the subsequent outputs from the same line.

**Review by villebro — APPROVED** [review]
Very nice!! A few nits and some general comments:
- should we checkout the currently checked out branch after the script completes?
- should we stash before checkout and later stash apply before the script completes?
- Could we create the headline like the current script does? The date would be the latest PR.

These are mostly improvement proposals, and we can probably work on refining this in subsequent PRs.

**Comment by dpgaspar** [comment]
@villebro thank you for the review. Addressed the following:

- checkout the currently checked out branch after the script completes
- create the headline like the current script does, The date is be the latest PR.

**Review by villebro — APPROVED** [review]
Nice! Good riddance `github-changes`! 🎉

**Pull request merged**